### PR TITLE
Support reading multipart data with `\n` (`LF`) lines

### PR DESCRIPTION
**Pull request by kxepal**
## What do these changes do?

While RFC clearly says about `CRLF` newlines, there quite a lot of
implementations which uses just `LF`. Even Python's stdlib produces
multiparts with `\n` newlines by default for compatibility reasons.

We wouldn't change how we produce multipart content - here we follow
RFC. However, we can detect `\n` lines quite easily which makes their
support quite cheap.

## Are there changes in behavior for the user?

No notable, except now they can read multipart data with "popular" newlines.

## Related issue number

#2302

### CODE REVIEW

**Comment by asvetlov** [comment]
Thanks for the PR!
Does newline style autodetection make sense?

**Review by thehesiod — CHANGES_REQUESTED** [review]
~~This does not fix `StreamReader.readline`, called via `MultipartReader.next`~~

**Comment by kxepal** [comment]
IIRC, on merge block right below the comments there should be some button about review cancellation. As backup plan, I can just change the latest commit hash via rebase over master or just ask @asvetlov for `sudo git merge`. In anyway, no worries about (:

**Comment by thehesiod** [comment]
ok so, after a few tries it works, will have to look into it more later, seems like a multithreading issue with aiohttp (one loop and server per thread with new release), but for now:

- I was calling `web.Request.multipart()`, however there's no param to set for the newline, so I now have to do: `MultipartReader(req.headers, req.content, newline=b'\n')`, should we add a param to this method?
- for automatic detection, this would be really nice because otherwise your server can only accept one type or the other, but not both.  Similar for reading multipart, somehow you'd have to detect/know what the newlines are otherwise.  I'd prefer having an auto-detect mechanism.

thanks guys! sorry about that little disruption, got my virtualenvs confused :)

**Comment by thehesiod** [comment]
btw I don't see anything about cancelling my review there, perhaps asvetlov or you can dismiss my review

**Comment by kxepal** [comment]
The initial idea is about to autodetect on MultipartReader side, while BodyPartReader have to be informed about what kind of newline to expect. So it should work in seconds case. But since the former one is deep in implementation details, everything should work for you without worry about what kind of newline is there.

**Comment by kxepal** [comment]
Hm...I should make a test case for this. Hold on.

**Comment by thehesiod** [comment]
sorry not following, from what I understand of this PR, you need to state upfront to your multipart reader what the newline is....which means if I write a server, two things are true:
1) It needs to know upfront what the multipart newline format is
2) It can only accept or format or another

These are two big negatives.  In my case it's ok, because I know upfront what it is, and I don't need to support mixed newline requests.  However in general this is not true if you have a generic server, you can't control what format clients will send their newlines.

**Comment by kxepal** [comment]
No, you don't need to care about what newline is. It handles by [this part](https://github.com/aio-libs/aiohttp/pull/3492/files#diff-235eeadd1960735e1191879aa500b7b7R648) of changes for you. In other words: first boundary dictates newline format fot the rest ones.

**Comment by thehesiod** [comment]
very cool! 🙇

**Comment by asvetlov** [comment]
Sorry, I don't follow the design.

Looks like `newline` argument is accepted here: https://github.com/aio-libs/aiohttp/pull/3492/files#diff-235eeadd1960735e1191879aa500b7b7R538 
but overridden later here: https://github.com/aio-libs/aiohttp/pull/3492/files#diff-235eeadd1960735e1191879aa500b7b7R651

Did I miss something?

**Comment by kxepal** [comment]
@asvetlov 
Multipart is recursive format. Once we detect newline at the very first boundary we have to pass it around for nested readers. So we have to have it as an argument and have to rewrite it during parsing, but only once. 
I just noticed that current implementation allows to have mix of different newlines - that's need to forbid for good.

**Comment by asvetlov** [comment]
Is it ready to merge?

**Comment by kxepal** [comment]
Sorry, not yet. Will be able to finish this a bit later today.

**Comment by kxepal** [comment]
@asvetlov 
Ok, I think it's ready. I was tries to catch the mixed newlines case when CRLF and LF may vary, but it seems to be not issue for us. PR rebased. If everythink ok for you and @thehesiod  let's merge it.

File: aiohttp/multipart.py [hunk 1]
```diff
@@ -237,11 +237,16 @@ class BodyPartReader:
...
 
-    def __init__(self, boundary: bytes,
-                 headers: Mapping[str, Optional[str]],
-                 content: StreamReader) -> None:
+    def __init__(
+        self,
+        boundary: bytes,
+        headers: Mapping[str, Optional[str]],
+        content: StreamReader,
+        newline: bytes = b'\r\n',
```
Comment (kxepal) [inline aiohttp/multipart.py]: To not cause any confusion, may be mark it somehow as internal? Like kwarg only or `_newline`. What do you think?

Comment (asvetlov) [reply aiohttp/multipart.py]: Makes sense for both keyword only and `_newline` proposals.
It will prevent users from using this *special* argument.

Comment (kxepal) [reply aiohttp/multipart.py]: Ok, will make both changes.

Comment (kxepal) [reply aiohttp/multipart.py]: Fixed in https://github.com/aio-libs/aiohttp/pull/3492/commits/5536671a48e9f72a4ff57f13c7a6ea895bbbd9fd

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
Thanks!

**Comment by thehesiod** [comment]
Nice work, thank you!!

**Comment by kxepal** [comment]
:beers:

File: CHANGES/2302.feature [hunk 2]
```diff
@@ -0,0 +1 @@
+Support reading multipart data with `\n` (`LF`) lines
```
Comment (webknjaz) [inline CHANGES/2302.feature:1]: Hm.. we need to watch for incorrect single quotes in RST...

**Pull request merged**